### ISSUE TITLE
Custom SchemaClient with fallback on default

### ISSUE THREAD

**Issue by haferwolle** [erosb/json-sKema#110]
Hi,

First of all, thanks for the great library.

To some extend, a similar question has been asked here: https://github.com/erosb/json-sKema/issues/94
But the solution doesn't work for me.

What I am trying to do:
I have JsonSchemas that have references to custom schemas, like "$ref":"http://example.xyz/Foo/Bar"

I'd like to be able to check if the URI starts with "http://example.xyz/", if so, I'd like to load the schema accordingly. 
In all other cases, the default SchemaClient should be used during validattion.

Here is my approach:
```
public class MySchemaClient implements SchemaClient {

    private SchemaClient defaultClient=SchemaLoaderConfig.createDefaultConfig(Map.of()).getSchemaClient();
    
    @NotNull
    @Override
    public InputStream get(@NotNull URI uri) {
        if (uri.toString().startsWith("http://example.xyz")) {
            String schemaJson = getCustomerSchema(uri);
            return new ByteArrayInputStream(schemaJson.getBytes(StandardCharsets.UTF_8));
        }
        
        // Fallback to the default SchemaClient if the URI does not match the custom prefix
        return defaultClient.get(uri);
    }

    @NotNull
    @Override
    public IJsonValue getParsed(@NotNull URI uri) {
        try {
            InputStream schemaStream = get(uri);
            return new JsonParser(schemaStream, uri).parse();
        } catch (Exception e) {
            throw new RuntimeException("Failed to parse schema from URI: " + uri, e);
        }
    }
}
```

For validation I then do:

```
SchemaLoaderConfig config=new SchemaLoaderConfig(new MySchemaClient(),""); //<-- I don't understand the concept of param InitialBaseURI, it seems to have no effect

Schema schema = new SchemaLoader(schemaForValidation, config).load();

ValidatorConfig validatorConfig=new ValidatorConfig(FormatValidationPolicy.ALWAYS);
Validator validator=Validator.create(schema,validatorConfig);

// running the validation
ValidationFailure result = validator.validate(jsonToValidate);
```

However, I get error, when MySchemaClient should resolve URIs that start with "mem://input", which seems to be a pseudo url for internally referenced components.

If I just use this config:

`SchemaLoaderConfig config = SchemaLoaderConfig.createDefaultConfig(Map.of());`

Then, all validations work, except, when I use one of the above mentioned "$ref", that points at a custom URI.

Long story short, I'd like to keep the default validation, but based on a URI prefix, I'd like to load my custom schemas. Is there any way to achieve this?

Thanks

**Comment by erosb** [comment]
Hello, 

would it maybe work for you to [pre-register your schemas](https://github.com/erosb/json-sKema?tab=readme-ov-file#pre-registering-schemas-by-uri-before-schema-loading) which are at http://example.xyz ?

Would that work for you with the current version?

I will try to put some work into cleaning up the `SchemaClient` internal interface, make it public and documented, but I can't promise any ETA on that.

> However, I get error, when MySchemaClient should resolve URIs that start with "mem://input", which seems to be a pseudo url for internally referenced components.

Yes, that's exactly the case, `mem://input` is the initial URI, used when the schema doesn't have a top-level ID. I will add this to the readme. See the [related section of the spec](https://json-schema.org/draft/2020-12/json-schema-core#name-initial-base-uri) for details.

**Comment by pancortexcom** [comment]
Hoi @erosb,

Thanks for your answer and sorry for my late response, but I figured it out two days after my initial question.
The solution was to not do this in MySchemaClient:

`private SchemaClient defaultClient=SchemaLoaderConfig.createDefaultConfig(Map.of()).getSchemaClient();

Instead, if I do outside MySchemaClient:

`
SchemaLoaderConfig defaultConfig = SchemaLoaderConfig.createDefaultConfig(Map.of());
new MySchemaClient(defaultConfig)
`

And then inside the constructor of MySchemaClient I do:
`this.defaultClient=defaultConfig.getSchemaClient()`

Then everything works as expected.

Not sure, what's the root cause for this behaviour. Might also be, that I am not aware of certain Kotlin specifics.
 Anyway, that way it works.
 
 In my case pre-registering the schemas was not an option.
 
 Thanks. your library is great.

**Comment by erosb** [comment]
Hello @haferwolle , this is now fixed in version `0.19.0`.
